mainnet-contracts/src/PufferVaultV3.sol

```diff
@@ -0,0 +1,217 @@
+     * @param stETHStrategy Address of the stETH strategy contract.
+     * @param eigenStrategyManager Address of the EigenLayer strategy manager contract.
+     * @param oracle Address of the PufferOracle contract.
+     * @param delegationManager Address of the delegation manager contract.
+     * @param bridgingConstructorParams Constructor parameters for bridging.
+     */
+    constructor(
+        IStETH stETH,
+        IWETH weth,
+        ILidoWithdrawalQueue lidoWithdrawalQueue,
+        IStrategy stETHStrategy,
+        IEigenLayer eigenStrategyManager,
+        IPufferOracle oracle,
+        IDelegationManager delegationManager,
+        BridgingConstructorParams memory bridgingConstructorParams
+    ) PufferVaultV2(stETH, weth, lidoWithdrawalQueue, stETHStrategy, eigenStrategyManager, oracle, delegationManager) {
+        _CONNEXT = IConnext(bridgingConstructorParams.connext);
+        XTOKEN = IERC20(bridgingConstructorParams.xToken);
+        _DESTINATION_DOMAIN = bridgingConstructorParams.destinationDomain;
+        LOCKBOX = IXERC20Lockbox(bridgingConstructorParams.lockBox);
+        L2_REWARD_MANAGER = bridgingConstructorParams.l2RewardManager;
+        _disableInitializers();
+    }
+
+    /**
+     * @notice Fallback function to receive ETH.
+     */
+    receive() external payable virtual override { }
+
+    /**
+     * @notice Returns the total assets held by the vault.
+     * @dev See {IERC4626-totalAssets}. pufETH, the shares of the vault, will be backed primarily by the WETH asset.
+     * However, at any point in time, the full backings may be a combination of stETH, WETH, Eigenpod ETH Rewards and ETH.
+     * `totalAssets()` is calculated by summing the following:
+     * - WETH held in the vault contract
+     * - ETH held in the vault contract
+     * - PUFFER_ORACLE.getLockedEthAmount(), which is the oracle-reported Puffer validator ETH locked in the Beacon chain
+     * - stETH held in the vault contract, in EigenLayer's stETH strategy, and in Lido's withdrawal queue. (we assume stETH is always 1:1 with ETH since it's rebasing)
+     * - ETH held in the eigenpods as a result of receiving rewards.
+     * NOTE on the native ETH deposits:
+     * When dealing with NATIVE ETH deposits, we need to deduct callvalue from the balance.
+     * The contract calculates the amount of shares (pufETH) to mint based on the total assets.
+     * When a user sends ETH, the msg.value is immediately added to address(this).balance.
+     * Since address(this.balance)` is used in calculating `totalAssets()`, we must deduct the `callvalue()` from the balance to prevent the user from minting excess shares.
+     * `msg.value` cannot be accessed from a view function, so we use assembly to get the callvalue.
+     * @return The total assets held by the vault.
+     */
+    function totalAssets() public view virtual override returns (uint256) {
+        uint256 callValue;
```

Comment:
this can probably by replaced by `return super.totalAssets() + getTotalRewardMintAmount();`